backend/controller/ingress/ingress.go

```diff
@@ -0,0 +1,233 @@
+package ingress
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/TBD54566975/ftl/backend/common/slices"
+	"github.com/TBD54566975/ftl/backend/controller/dal"
+	"github.com/TBD54566975/ftl/backend/schema"
+)
+
+type path []string
+
+func (p path) String() string {
+	return strings.TrimLeft(strings.Join(p, ""), ".")
+}
+
+func GetIngressRoute(routes []dal.IngressRoute, method string, path string) (*dal.IngressRoute, error) {
+	var matchedRoutes = slices.Filter(routes, func(route dal.IngressRoute) bool {
+		return matchAndExtractAllSegments(route.Path, path, func(segment, value string) {})
+	})
+
+	if len(matchedRoutes) == 0 {
+		return nil, dal.ErrNotFound
+	}
+
+	// TODO: add load balancing at some point
+	route := matchedRoutes[rand.Intn(len(matchedRoutes))] //nolint:gosec
+	return &route, nil
+}
+
+func matchAndExtractAllSegments(pattern, urlPath string, onMatch func(segment, value string)) bool {
```

Comment:
Should this be called something different? Like just `matchSegments()`?

Reply:
I like that!